<a href="https://colab.research.google.com/github/mani2222278/Rebounce-Applied-AI-and-Analytics-code/blob/main/python_301.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MP-3 Clean ,Analyse $Summarise Real Dataset

# Import Pandas $ NumPy



In [3]:
import pandas as pd
import numpy as np

#Dataset Load

In [4]:
url = "https://huggingface.co/datasets/aarav912/online-retail/resolve/main/online_retail.csv"
df = pd.read_csv(url)

print("Dataset loaded successfully.")
print("Shape:", df.shape)
df.head()


Dataset loaded successfully.
Shape: (541909, 8)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [6]:
df.dtypes

,0
InvoiceNo,object
StockCode,object
Description,object
Quantity,int64
InvoiceDate,object
UnitPrice,float64
CustomerID,float64
Country,object


In [7]:
df.isna().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


In [8]:
df.duplicated().sum()

np.int64(5268)

In [9]:
df["Sales"] = df["Quantity"] * df["UnitPrice"]

print("Sales column created successfully.")
print(df[["Quantity", "UnitPrice", "Sales"]].head())

Sales column created successfully.
   Quantity  UnitPrice  Sales
0         6       2.55  15.30
1         6       3.39  20.34
2         8       2.75  22.00
3         6       3.39  20.34
4         6       3.39  20.34


In [10]:
median_sales = np.median(df["Sales"])

print("Median Sales:", median_sales)

Median Sales: 9.75


In [11]:
median_sales = np.median(df["Sales"].to_numpy())

print("Median Sales:", median_sales)

Median Sales: 9.75


#PART 2-CLEAN THE DATA

In [28]:
df = df.dropna(subset=["CustomerID"])

print("Shape after removing missing CustomerID:", df.shape)

Shape after removing missing CustomerID: (392692, 9)


Reason:
I removed rows with missing CustomerID becouse a valid customer number is needed for customer-level analysis

#convert InvoiceDate

In [13]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

print("Shape after converting InvoiceDate:", df.shape)
print("InvoiceDate data type:", df["InvoiceDate"].dtype)

Shape after converting InvoiceDate: (406829, 9)
InvoiceDate data type: datetime64[ns]


Remove duplicates

In [14]:
df = df.drop_duplicates()

print("Shape after removing duplicates:", df.shape)

Shape after removing duplicates: (401604, 9)


Remove Invalid transaxtion

In [15]:
df = df[(df["Quantity"] > 0) & (df["UnitPrice"] > 0)]

print("Shape after removing invalid transactions:", df.shape)

Shape after removing invalid transactions: (392692, 9)


Sales column

In [16]:
df["Sales"] = df["Quantity"] * df["UnitPrice"]

print("Sales column created successfully.")
print(df[["Quantity", "UnitPrice", "Sales"]].head())

Sales column created successfully.
   Quantity  UnitPrice  Sales
0         6       2.55  15.30
1         6       3.39  20.34
2         8       2.75  22.00
3         6       3.39  20.34
4         6       3.39  20.34


 Part 3 - Numpy Analysis

In [17]:

mean_sales = np.mean(df["Sales"])
print("Mean Sales:", mean_sales)

median_sales = np.median(df["Sales"])
print("Median Sales:", median_sales)

min_sales = np.min(df["Sales"])
print("Minimum Sales:", min_sales)

max_sales = np.max(df["Sales"])
print("Maximum Sales:", max_sales)

std_sales = np.std(df["Sales"])
print("Standard Deviation:", std_sales)

Mean Sales: 22.6314997351614
Median Sales: 12.45
Minimum Sales: 0.001
Maximum Sales: 168469.6
Standard Deviation: 311.09882822360305


Explanation:
print("The mean and median are different, which shows that the sales values are not evenly distributed.")
print("If the mean is higher than the median, a small number of very large orders may be pulling the average upward.")

 Part 4 - Pandas Filtering

In [18]:

sales_above_500 = df[df["Sales"] > 500]

print("Number of transactions with Sales above 500:", len(sales_above_500))
print(sales_above_500)

Number of transactions with Sales above 500: 913
       InvoiceNo StockCode                     Description  Quantity  \
178       536387     79321                   CHILLI LIGHTS       192   
179       536387     22780  LIGHT GARLAND BUTTERFILES PINK       192   
180       536387     22779      WOODEN OWLS LIGHT GARLAND        192   
181       536387     22466   FAIRY TALE COTTAGE NIGHTLIGHT       432   
182       536387     21731   RED TOADSTOOL LED NIGHT LIGHT       432   
...          ...       ...                             ...       ...   
540069    581457     23401   RUSTIC MIRROR WITH LACE HEART       126   
540070    581458     22197                  POPCORN HOLDER      1500   
540071    581459     22197                  POPCORN HOLDER      1200   
540421    581483     23843     PAPER CRAFT , LITTLE BIRDIE     80995   
540425    581485     20749      ASSORTED COLOUR MINI CASES        84   

               InvoiceDate  UnitPrice  CustomerID         Country      Sales  
178    

Largest transaction

In [19]:

largest_10 = df.sort_values("Sales", ascending=False).head(10)

print("10 Largest Transactions:")
print(largest_10)

10 Largest Transactions:
       InvoiceNo StockCode                          Description  Quantity  \
540421    581483     23843          PAPER CRAFT , LITTLE BIRDIE     80995   
61619     541431     23166       MEDIUM CERAMIC TOP STORAGE JAR     74215   
222680    556444     22502       PICNIC BASKET WICKER 60 PIECES        60   
173382    551697      POST                              POSTAGE         1   
348325    567423     23243  SET OF TEA COFFEE SUGAR TINS PANTRY      1412   
160546    550461     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
52711     540815     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114   
421601    573003     23084                   RABBIT NIGHT LIGHT      2400   
52709     540815    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   
160542    550461    85123A   WHITE HANGING HEART T-LIGHT HOLDER      1930   

               InvoiceDate  UnitPrice  CustomerID         Country      Sales  
540421 2011-12-09 09:15:00       2.08     16446.

Germany transaction

In [20]:
germany = df[df["Country"] == "Germany"]

print("Number of transactions from Germany:", len(germany))
print(germany)

Number of transactions from Germany: 9025
       InvoiceNo StockCode                          Description  Quantity  \
1109      536527     22809              SET OF 6 T-LIGHTS SANTA         6   
1110      536527     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         6   
1111      536527     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        12   
1112      536527     22242        5 HOOK HANGER MAGIC TOADSTOOL        12   
1113      536527     22244           3 HOOK HANGER MAGIC GARDEN        12   
...          ...       ...                                  ...       ...   
541801    581578     22993         SET OF 4 PANTRY JELLY MOULDS        12   
541802    581578     22907     PACK OF 20 NAPKINS PANTRY DESIGN        12   
541803    581578     22908        PACK OF 20 NAPKINS RED APPLES        12   
541804    581578     23215     JINGLE BELL HEART ANTIQUE SILVER        12   
541805    581578     22736          RIBBON REEL MAKING SNOWMEN         10   

               InvoiceDate  UnitP

Quantity above 1000

In [36]:
quantity_above_1000 = df[df["Quantity"] > 1000]

print("Transactions with Quantity above 1000:")
print(quantity_above_1000)

Transactions with Quantity above 1000:
       InvoiceNo StockCode                        Description  Quantity  \
4850      536809     84950     ASSORTED COLOUR T-LIGHT HOLDER      1824   
4945      536830     84077  WORLD WAR 2 GLIDERS ASSTD DESIGNS      2880   
4946      536830     21915             RED  HARMONICA IN BOX       1400   
6365      536890    17084R              ASSORTED INCENSE PACK      1440   
16435     537659     22189            CREAM HEART CARD HOLDER      1008   
...          ...       ...                                ...       ...   
533812    581115     22413    METAL SIGN TAKE IT OR LEAVE IT       1404   
534952    581175     23084                 RABBIT NIGHT LIGHT      1440   
540070    581458     22197                     POPCORN HOLDER      1500   
540071    581459     22197                     POPCORN HOLDER      1200   
540421    581483     23843        PAPER CRAFT , LITTLE BIRDIE     80995   

               InvoiceDate  UnitPrice  CustomerID         Co

 PART5 - GroupBy  and aggregation

In [22]:
summary = df.groupby("Country").agg(
    Number_of_Transactions=("Sales", "count"),
    Total_Sales=("Sales", "sum"),
    Average_Sales=("Sales", "mean")
).reset_index()

print("Country Summary:")
print(summary)

Country Summary:
                 Country  Number_of_Transactions  Total_Sales  Average_Sales
0              Australia                    1181   138453.810     117.234386
1                Austria                     398    10198.680      25.624824
2                Bahrain                      17      548.400      32.258824
3                Belgium                    2031    41196.340      20.283772
4                 Brazil                      32     1143.600      35.737500
5                 Canada                     151     3666.380      24.280662
6        Channel Islands                     747    20440.540      27.363507
7                 Cyprus                     603    13502.850      22.392786
8         Czech Republic                      25      826.740      33.069600
9                Denmark                     380    18955.340      49.882474
10                  EIRE                    7226   265262.460      36.709446
11    European Community                      60     1300.2

In [24]:
top_5_countries = summary.sort_values(
    "Total_Sales", ascending=False
).head(5)

print("Top 5 Countries by Total Sales:")
print(top_5_countries)

Top 5 Countries by Total Sales:
           Country  Number_of_Transactions  Total_Sales  Average_Sales
35  United Kingdom                  349203  7285024.644      20.861862
23     Netherlands                    2359   285446.340     121.003111
10            EIRE                    7226   265262.460      36.709446
14         Germany                    9025   228678.400      25.338327
13          France                    8326   208934.310      25.094200


In [25]:
highest_average = summary.sort_values(
    "Average_Sales", ascending=False
).head(1)

print("Country with Highest Average Sales:")
print(highest_average)

Country with Highest Average Sales:
        Country  Number_of_Transactions  Total_Sales  Average_Sales
23  Netherlands                    2359    285446.34     121.003111


In [26]:
top_country = summary.sort_values(
    "Total_Sales", ascending=False
).iloc[0]

print("Top-selling Country:", top_country["Country"])
print("Number of Transactions:", top_country["Number_of_Transactions"])
print("Total Sales:", top_country["Total_Sales"])

Top-selling Country: United Kingdom
Number of Transactions: 349203
Total Sales: 7285024.644


In [27]:
summary = summary.sort_values(
    "Total_Sales",
    ascending=False
)

print("Final Country Summary:")
print(summary)

Final Country Summary:
                 Country  Number_of_Transactions  Total_Sales  Average_Sales
35        United Kingdom                  349203  7285024.644      20.861862
23           Netherlands                    2359   285446.340     121.003111
10                  EIRE                    7226   265262.460      36.709446
14               Germany                    9025   228678.400      25.338327
13                France                    8326   208934.310      25.094200
0              Australia                    1181   138453.810     117.234386
30                 Spain                    2479    61558.560      24.832013
32           Switzerland                    1841    56443.950      30.659397
3                Belgium                    2031    41196.340      20.283772
31                Sweden                     450    38367.830      85.261844
19                 Japan                     321    37416.370     116.561900
24                Norway                    1071    3

 PART 6 -Merge and region analysis

In [29]:
region_lookup = {
    "United Kingdom": "Europe",
    "France": "Europe",
    "Germany": "Europe",
    "Spain": "Europe",
    "Netherlands": "Europe",
    "Belgium": "Europe",
    "Switzerland": "Europe",
    "Portugal": "Europe",
    "Australia": "Oceania",
    "Japan": "Asia",
    "USA": "Americas",
}

region_df = pd.DataFrame(list(region_lookup.items()), columns=["Country", "Region"])

# Left merge
merged = summary.merge(region_df, on="Country", how="left")
merged["Region"] = merged["Region"].fillna("Other")

print(merged.head(15))

# Total sales per region
region_summary = (
    merged.groupby("Region")["Total_Sales"]
          .sum()
          .sort_values(ascending=False)
)
print("\n--- Total Sales by Region ---")
print(region_summary)

           Country  Number_of_Transactions  Total_Sales  Average_Sales  \
0   United Kingdom                  349203  7285024.644      20.861862   
1      Netherlands                    2359   285446.340     121.003111   
2             EIRE                    7226   265262.460      36.709446   
3          Germany                    9025   228678.400      25.338327   
4           France                    8326   208934.310      25.094200   
5        Australia                    1181   138453.810     117.234386   
6            Spain                    2479    61558.560      24.832013   
7      Switzerland                    1841    56443.950      30.659397   
8          Belgium                    2031    41196.340      20.283772   
9           Sweden                     450    38367.830      85.261844   
10           Japan                     321    37416.370     116.561900   
11          Norway                    1071    36165.440      33.767918   
12        Portugal                    

In [31]:
summary_region = pd.merge(
    summary,
    region_df,
    on="Country",
    how="left"
)

summary_region["Region"] = summary_region["Region"].fillna("Other")

print("Country Summary with Region:")
print(summary_region)

Country Summary with Region:
                 Country  Number_of_Transactions  Total_Sales  Average_Sales  \
0         United Kingdom                  349203  7285024.644      20.861862   
1            Netherlands                    2359   285446.340     121.003111   
2                   EIRE                    7226   265262.460      36.709446   
3                Germany                    9025   228678.400      25.338327   
4                 France                    8326   208934.310      25.094200   
5              Australia                    1181   138453.810     117.234386   
6                  Spain                    2479    61558.560      24.832013   
7            Switzerland                    1841    56443.950      30.659397   
8                Belgium                    2031    41196.340      20.283772   
9                 Sweden                     450    38367.830      85.261844   
10                 Japan                     321    37416.370     116.561900   
11         

In [32]:
region_sales = summary_region.groupby("Region")["Total_Sales"].sum().reset_index()

region_sales = region_sales.sort_values(
    "Total_Sales",
    ascending=False
)

print("Total Sales by Region:")
print(region_sales)

Total Sales by Region:
     Region  Total_Sales
2    Europe  8200658.384
4     Other   507099.940
3   Oceania   138453.810
1      Asia    37416.370
0  Americas     3580.390


In [33]:
top_region = region_sales.iloc[0]

print("Region with Highest Total Sales:", top_region["Region"])
print("Total Sales:", top_region["Total_Sales"])

Region with Highest Total Sales: Europe
Total Sales: 8200658.384000001


PART 7 – ANSWER THE QUESTIONS

1. Which country generated the highest total sales?
ANSWAR - The United Kingdom generated the highest total sales, with approximately £7,285,024.64 in total sales.
2. Which country had the highest average transaction value?
ANSWAR - The Netherlands had the highest average transaction value, with an average of approximately £121.00 per transaction.
3. Which region performed best?
ANSWAR - The Europe region performed the best because it generated the highest total sales of approximately £8.20 million.
4. What is one interesting pattern you noticed in the data?
ANSWAR - An interesting pattern is that the United Kingdom dominates total sales and transaction count, while the Netherlands has a much higher average transaction value.
5. What is one data-quality problem you ran into, and how did you handle it?ANSWAR - The dataset contained missing CustomerID values and transactions with zero or negative Quantity/UnitPrice. I removed rows with missing CustomerID and filtered the data t

Part 8-final insight





# The cleaned Online Retail dataset shows that the United Kingdom dominates
# total sales, contributing more than £7.2 million. However, the average
# transaction value is highest in the Netherlands (£121), followed by
# Australia and Japan. This comparison reveals that while the UK generates
# volume, several smaller markets generate higher-value wholesale orders.
#
# The Sales distribution is right-skewed: the mean (£22.63) sits well above
# the median (£12.45), confirming that a small number of very large orders
# pull the average upward.
#
# One limitation is that we dropped all rows with missing CustomerID and
# all cancellation/refund rows. The final numbers therefore reflect only
# completed sales linked to known customers, and may understate the true
# scale of the business if many guest checkouts occurred